rfd/0163-vnet.md

```diff
@@ -0,0 +1,971 @@
+gVisor "gvisor.dev/gvisor/pkg/tcpip/stack", which is a Go library available
+under the Apache 2.0 license.
+
+When VNet accepts a TCP connection on an address assigned to an app, it will
+proxy the connection to the app over an authenticated tunnel, in much the same
+way that `tsh proxy app` works today.
+
+### Protocols
+
+VNet will initially support TCP and HTTP application access.
+
+For TCP apps (any app with a `spec.uri` starting with `tcp://`) VNet will proxy
+all TCP connections directly from clients to the application unaltered.
+If the destination app is a TLS server, the client may connect to it over TLS.
+Any custom protocol over TCP is supported.
+
+HTTP apps (any app with a `spec.uri` starting with `http://` or `https://`) will be
+accessible through VNet over plain HTTP.
+They will also continue to be accessible over HTTPS from the browser as they are
+today - to support this VNet will detect TLS connections to HTTP app addresses
+and forward TCP directly to the Teleport Proxy instead of opening a tunnel all
+the way to the app, with authentication and routing handled by the proxy via
+cookies sent by the web browser.
+Teleport HTTP app access enforces that all proxied connections are actually HTTP
+and supports custom header rewriting, this will be the same when connecting
+through VNet.
+
+We could consider UDP support, but app access does not support UDP at all right
+now, so it will be out of scope for this RFD.
+
+We can explore the possibility of supporting Kubernetes and Database access in
+future versions, but these protocols come with additional challenges.
+We would need different certs to access a single database depending on the db
+username, database name, etc.
+Would we expose each at a different domain name?
+There may be similar challenges for kubernetes user and group impersonation.
+Currently these parameters are accepted at the `tsh proxy` command but VNet
+has no similar mechanism available.
+
+### Domain name assignment
+
+When resolving DNS queries, VNet will search for an app with `spec.public_addr`
+exactly matching the name in the query.
+If a matching app is found, it will be assigned an IP address in the VNet IP
+range and the DNS query will resolve to that address.
+
+The default value for an app's `spec.public_addr` is `<app-name>.<proxy-public-address>`,
+but that can be overridden with any value by a cluster admin.
+For an app named `api` in a Teleport cluster at `teleport.example.com` this
+would look like `api.teleport.example.com` by default.
+This is the same address that applications are already accessible at in the web
+today.
+
+Because we will use split DNS to configure VNet as a DNS server only for
+specific subdomains, if an overridden `spec.public_addr` is NOT suffixed with
+the proxy's public address an admin will need to configure a custom DNS zone in
+the cluster, see [custom DNS zones](#custom-dns-zones).
+
+### Custom DNS zones
```

Comment:
Since a custom DNS zone may not be fully controlled by Teleport, will we validate that hosts are not already registered in DNS before we allow them to be registered in Teleport?  Or what thought have we given towards Teleport potentially being leveraged for a DNS rebinding attack?

One immediate suggestion for improvement is to require validation from the DNS controller.  It should not be possible to register a domain in Teleport that the domain is not confirming it should be accessible for binding.

Possibly requiring a DNS TXT record could be a good way for the proxy to validate the zone is available for use.

Reply:
Hmm, I was thinking of VNet "hijacking" an existing DNS name as a good migration path. DNS entries for all existing app `public_addrs` are actually expected to already exist. Since this DNS server is only local on a single machine, I thought it was fine.

I guess we are concerned about a scenario where an attacker adds an app to the Teleport cluster with `public_addr` overriding a DNS name that other VNet users in the same Teleport cluster already connect to without teleport and without HTTPS, and those VNet users will now start connecting to the teleport app instead.

I feel like requiring public DNS records would be a pretty big configuration burden for VNet adopters. The possible DNS names that could be overridden in VNet will already be limited to the configured custom DNS zones, do you think it could be sufficient to make that cluster level configuration of custom DNS zones admin-protected within Teleport?

Reply:
I think we should put significant consideration here.  Although requiring admin level access to help configure these zones is helpful, I still think there is significant risk here.

Say for example a compromised admin account, or internal actor who wanted to conduct an attack, and knew the organization used `okta`.  What would prevent them from registering `okta.com` as a custom DNS zone, with an application that matches the organization name and hosting a malicious fake Okta site?

Now all employees using this feature could easily accidentally load a fake Okta and allow an attacker a significant ability to expand their attack (granted, in this example it would be more challenging with FIDO, but very doable with OTP MFA).

My concern is that this may make Teleport a potential weapon that can be used for dns rebinding attacks beyond just the domains controlled by the Teleport cluster.  The best way to prevent this type of attack is ensuring that the domain agrees to be able to be controlled by our DNS.

An alternative could be we try not to overload any DNS which is available in public DNS records.  But I was concerned this may be more error prone (particularly if a customer is trying to transition as described in the RFD).

Reply:
Teleport and VNet shouldn't be able to get a trusted TLS cert for `okta.com`

Now, since the current written design provides plain HTTP access to apps, it could be a concern that someone would open http://okta.com. HSTS should help here to some extent

One option I'm thinking of would be to only override public DNS names where they already resolve to the teleport proxy address. Unfortunately, there may be many proxy address and they could be changing. I guess we could actually connect to that address over TLS and be able to tell whether it is the proxy somehow

Reply:
I think getting approval from the domain controller is the good citizen behavior here, even if we don't do https termination (which we may add in the future anyways).  A DNS txt record seems like an easy check, but checking for resolution in the public DNS seems acceptable as well.

Reply:
I like the idea of a TXT record. A little more work to set up, but I think that's okay. Most people probably won't need custom DNS zones.

Reply:
I added a requirement for a TXT record for custom DNS zones to the RFD.